---
title: "GSFA Results of LUHMES CROP-seq Data"
subtitle: "-- for manuscript"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m")`'
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(WebGestaltR)

# set default chunk output
knitr::opts_chunk$set(echo = T,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

# Background

## Input data and pre-processing

Source:   
[High-throughput single-cell functional elucidation of neurodevelopmental disease-associated genes reveals convergent mechanisms altering neuronal differentiation](https://genome.cshlp.org/content/30/9/1317.full), GEO accession: GSE142078.

Perturbations:    
CRISPR knock-down of 14 autism spectrum disorder (ASD)â€“associated genes (3 gRNAs per gene) + 5 non-targeting gRNAs.

Cells:    
Lund human mesencephalic (LUHMES) neural progenitor cell line.   
Cells from 3 batches were merged together into 1 analysis. All cells have only a single type of gRNA readout. Quality control resulted in 8708 cells.

Genes:    
**Top 6000 genes ranked by the multinomial deviance statistics were kept.**

Normalization:    
Deviance residual transformation.   
**Batch effect, unique UMI count, library size, and mitochondria percentage were all corrected for. The corrected and scaled expression data were used as input for subsequent GSFA.**

## GSFA

Input:    
In addition to the expression matrix, we have a binary perturbation matrix ($G$) consisting of 15 types (14 genes + negative control) of gene-level knock-down conditions across cells.

For Gibbs sampling, we specified 20 factors initialized from truncated SVD. Results were obtained from 3000 iterations, with the posterior mean estimates averaged over the last 1000 iterations.

```{r}
## Load data and plotting functions
source("../R/plotting_functions.R")
data_folder <- "../data/LUHMES/"

## Load the lighter GSFA output from "R/run_LUHMES.R":
fit <- readRDS(paste0(data_folder, "gsfa_fit.light.rds"))
gibbs_PM <- fit$posterior_means
lfsr_mat <- fit$lfsr[, -ncol(fit$lfsr)]
G_mat <- readRDS(paste0(data_folder, "perturbation_matrix.rds"))
genes_df <- readRDS(paste0(data_folder, "top6k_genes.rds"))
interest_df <- readRDS(paste0(data_folder, "selected_neuronal_markers.rds"))
KO_names <- colnames(lfsr_mat)
```

# Factor ~ Perturbation Association

## Perturbation effects on factors ($\beta$)

All targets and factors (Figure S6A):
```{r Fig_S6A, fig.width=9, fig.height=5.5}
dotplot_beta_PIP(t(gibbs_PM$Gamma_pm), t(gibbs_PM$beta_pm),
                 marker_names = KO_names,
                 reorder_markers = c(KO_names[KO_names!="Nontargeting"], "Nontargeting"),
                 inverse_factors = F) +
  coord_flip()
```

```{r fig.width=8, fig.height=5.5, eval=F}
## Similar visualization using GSFA built-in functions:
GSFA::dotplot_beta_PIP(fit,
                       target_names = KO_names,
                       reorder_targets = c(KO_names[KO_names!="Nontargeting"], "Nontargeting"))
```

Effects of selected perturbations on factors (Figure 5A):

```{r Fig_5A, fig.width=4, fig.height=4}
targets <- c("ADNP", "ARID1B", "ASH1L", "CHD2", "DYRK1A", "PTEN", "SETD5")
complexplot_perturbation_factor(gibbs_PM$Gamma_pm[-nrow(gibbs_PM$Gamma_pm), ],
                                gibbs_PM$beta_pm[-nrow(gibbs_PM$beta_pm), ],
                                marker_names = KO_names,
                                reorder_markers = targets,
                                reorder_factors = c(4, 9, 16))
```

```{r fig.width=4, fig.height=4, eval=F}
## Similar visualization using GSFA built-in functions:
GSFA::dotplot_beta_PIP(fit,
                       target_names = KO_names,
                       reorder_targets = targets, reorder_factors = c(4, 9, 16))
```

## Factor-perturbation association p values
```{r fig.width=10, fig.height=5}
gibbs_res_tb <- make_gibbs_res_tb(gibbs_PM, G_mat, compute_pve = F)
heatmap_matrix <- gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(G_mat)

summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA\n(15 Targets x 20 Factors)")
```

# Factor Interpretation

## Correlation within factors

```{r fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Pairwise correlation within factors (Z)",
                           label_size = 10)
```

```{r fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = (gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Pairwise correlation within \nbinarized gene loadings (F_pm > 0.95)",
                           label_size = 10)
```

## Gene loading in factors

Neuronal differentiation markers:

```{r echo=FALSE}
interest_df <- interest_df[interest_df$gene_name %in% genes_df$Name, ]
rownames(interest_df) <- NULL
knitr::kable(interest_df) %>%
    kable_styling() %>% scroll_box(width = '100%', height = '400px')
```

All factors (Figure S6B):

```{r Fig_S6B, fig.width=10, fig.height=5}
complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm)
```

Selected factors (Figure 5C):

```{r Fig_5C, fig.width=6.5, fig.height=5}
complexplot_gene_factor(genes_df, interest_df, gibbs_PM$F_pm, gibbs_PM$W_pm,
                        reorder_factors = c(4, 9, 16))
```

## GO enrichment analysis in factors

Foreground genes: Genes w/ non-zero loadings in each factor (gene PIP > 0.95);    
Background genes: all 6000 genes used in GSFA;    
Statistical test: hypergeometric test (over-representation test);  
Gene sets: Gene ontology "Biological Process" (non-redundant).

```{r eval=FALSE}
## The "WebGestaltR" tool needs Internet connection.
enrich_db <- "geneontology_Biological_Process_noRedundant"
PIP_mat <- gibbs_PM$F_pm
enrich_res_by_factor <- list()
for (i in 1:ncol(PIP_mat)){
  enrich_res_by_factor[[i]] <- 
    WebGestaltR::WebGestaltR(enrichMethod = "ORA",
                             organism = "hsapiens",
                             enrichDatabase = enrich_db,
                             interestGene = genes_df[PIP_mat[, i] > 0.05, ]$ID,
                             interestGeneType = "ensembl_gene_id",
                             referenceGene = genes_df$ID,
                             referenceGeneType = "ensembl_gene_id",
                             isOutput = F)
}
```

```{r echo=FALSE}
enrich_res_by_factor <- readRDS(paste0(data_folder, "GO_enrich_in_GSFA_factors.rds"))
```

Figure 5D:

```{r Fig_5D_factor4, fig.height=2.4, fig.width=7}
factor_indx <- 4
terms_of_interest <- c("regulation of ion transmembrane transport",
                       "regulation of trans-synaptic signaling",
                       "axon development",
                       "regulation of neuron projection development")
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r Fig_5D_factor9, fig.height=2.4, fig.width=7}
factor_indx <- 9
terms_of_interest <- c("actin filament organization",
                       "cell fate commitment",
                       "regulation of neuron projection development",
                       "regulation of cell morphogenesis")
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r Fig_5D_factor16, fig.height=1.8, fig.width=7}
factor_indx <- 16
terms_of_interest <- c("developmental growth involved in morphogenesis",
                       "axon development")
barplot_top_enrich_terms(enrich_res_by_factor[[factor_indx]],
                         terms_of_interest = terms_of_interest, 
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

# DEG Interpretation

## Number of DEGs detected by different methods

```{r}
fdr_cutoff <- 0.05
lfsr_cutoff <- 0.05
```

```{r echo=FALSE}
lfsr_signif_num <- colSums(lfsr_mat < lfsr_cutoff)
signif_num_tb <- t(data.frame(KO = names(lfsr_signif_num),
                           Num_genes = lfsr_signif_num,
                           row.names = NULL))
knitr::kable(rbind(signif_num_tb[, 1:5], signif_num_tb[, 6:10], signif_num_tb[, 11:15]),
             caption = "Number of DEGs detected by GSFA:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

```{r deseq}
deseq_list <- readRDS(paste0(data_folder, "DE_result_DESeq2.rds"))
deseq_signif_counts <- sapply(deseq_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r mast}
mast_list <- readRDS(paste0(data_folder, "DE_result_MAST.rds"))
mast_signif_counts <- sapply(mast_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r}
scmageck_res <- readRDS(paste0(data_folder, "DE_result_scMAGeCK.rds"))
colnames(scmageck_res$fdr)[colnames(scmageck_res$fdr) == "NegCtrl"] <- "Nontargeting"
scmageck_signif_counts <- colSums(scmageck_res$fdr[, KO_names] < fdr_cutoff)
```

```{r dge_comparison_merge}
dge_comparison_df <- data.frame(Perturbation = names(lfsr_signif_num),
                                GSFA = lfsr_signif_num,
                                scMAGeCK = scmageck_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_signif_counts)
dge_comparison_df$Perturbation[dge_comparison_df$Perturbation == "Nontargeting"] <- "NegCtrl"
```

Figure 5E:

```{r Fig_5E, fig.width=13, fig.height=5}
dge_plot_df <- reshape2::melt(dge_comparison_df, id.var = "Perturbation",
                              variable.name = "Method", value.name = "Num_DEGs")
dge_plot_df$Perturbation <- factor(dge_plot_df$Perturbation,
                                   levels = c("NegCtrl", KO_names[KO_names!="Nontargeting"]))
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Target gene",
       y = "Number of DEGs",
       title = "Number of DEGs detected by different methods") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

## Perturbation effects on marker genes

### GSFA

Figure 5G:

```{r Fig_5G, fig.width=6.5, fig.height=5}
targets <- c("ADNP", "ARID1B", "ASH1L", "CHD2", "DYRK1A", "PTEN", "SETD5")
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = lfsr_mat,
                              effect_mat = gibbs_PM$W_pm %*%
                                t(gibbs_PM$beta_pm[-nrow(gibbs_PM$beta_pm), ]))
```

```{r fig.width=5.5, fig.height=5, eval=F}
## Similar visualization using GSFA built-in functions:
GSFA::dotplot_total_effect(fit,
                           gene_indices = match(interest_df$gene_ID, rownames(lfsr_mat)),
                           gene_names = interest_df$gene_name,
                           reorder_targets = targets, 
                           plot_max_score = 0.2)
```

### scMAGeCK

Figure 5H:

```{r Fig_5H, fig.width=6.5, fig.height=5}
score_mat <- scmageck_res$score
fdr_mat <- scmageck_res$fdr
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = "FDR",
                              effect_mat = score_mat, effect_name = "scMAGeCK\nselection score", 
                              score_break = c(-0.2, 0, 0.2),
                              color_break = c("blue3", "grey90", "red3"))
```

### DESeq2

```{r}
FC_mat <- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) <- interest_df$gene_name
colnames(FC_mat) <- targets
fdr_mat <- FC_mat
for (m in targets){
  FC_mat[, m] <- deseq_list[[m]]$log2FoldChange[match(interest_df$gene_ID,
                                                      deseq_list[[m]]$geneID)]
  fdr_mat[, m] <- deseq_list[[m]]$FDR[match(interest_df$gene_ID, deseq_list[[m]]$geneID)]
}
```

Figure S5C:

```{r Fig_S5C, fig.width=6.5, fig.height=5}
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = "FDR",
                              effect_mat = FC_mat, effect_name = "DESeq2 log2FC",
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c("blue3", "grey90", "red3"))
```

### MAST

```{r}
FC_mat <- matrix(nrow = nrow(interest_df), ncol = length(targets))
rownames(FC_mat) <- interest_df$gene_name
colnames(FC_mat) <- targets
fdr_mat <- FC_mat
for (m in targets){
  FC_mat[, m] <- mast_list[[m]]$logFC[match(interest_df$gene_ID, mast_list[[m]]$geneID)]
  fdr_mat[, m] <- mast_list[[m]]$FDR[match(interest_df$gene_ID, mast_list[[m]]$geneID)]
}
```

Figure S5D:

```{r Fig_S5D, fig.width=6.5, fig.height=5}
complexplot_gene_perturbation(genes_df, interest_df,
                              targets = targets,
                              lfsr_mat = fdr_mat, lfsr_name = "FDR",
                              effect_mat = FC_mat, effect_name = "MAST logFC",
                              score_break = c(-0.4, 0, 0.4),
                              color_break = c("blue3", "grey90", "red3"))
```

## GO enrichment in DEGs

Foreground genes: Genes w/ GSFA LFSR < 0.05 under each perturbation;    
Background genes: all 6000 genes used in GSFA;    
Statistical test: hypergeometric test (over-representation test);    
Gene sets: Gene ontology "Biological Process" (non-redundant).

```{r eval=FALSE}
## The "WebGestaltR" tool needs Internet connection.
targets <- names(lfsr_signif_num)[lfsr_signif_num > 0]
enrich_db <- "geneontology_Biological_Process_noRedundant"
enrich_res <- list()
for (i in targets){
  print(i)
  interest_genes <- genes_df %>% mutate(lfsr = lfsr_mat[, i]) %>%
    filter(lfsr < lfsr_cutoff) %>% pull(ID)
  enrich_res[[i]] <- 
    WebGestaltR::WebGestaltR(enrichMethod = "ORA",
                             organism = "hsapiens",
                             enrichDatabase = enrich_db,
                             interestGene = interest_genes,
                             interestGeneType = "ensembl_gene_id",
                             referenceGene = genes_df$ID,
                             referenceGeneType = "ensembl_gene_id",
                             isOutput = F)
}
```

```{r echo=FALSE}
enrich_res <- readRDS(paste0(data_folder, "GO_enrich_in_GSFA_DEGs.rds"))
```

```{r}
signif_GO_list <- list()
for (i in names(enrich_res)) {
  signif_GO_list[[i]] <- enrich_res[[i]] %>%
    dplyr::filter(FDR < 0.05) %>%
    dplyr::select(geneSet, description, size, enrichmentRatio, pValue) %>%
    mutate(target = i)
}
signif_term_df <- do.call(rbind, signif_GO_list) %>%
  group_by(geneSet, description, size) %>%
  summarise(pValue = min(pValue)) %>%
  ungroup()

abs_FC_colormap <- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c("grey95", "#77d183", "#255566"))
```

```{r}
targets <- names(enrich_res)
enrich_table <- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) <- targets
for (i in 1:ncol(enrich_table)){
  m <- colnames(enrich_table)[i]
  enrich_df <- enrich_res[[m]] %>% dplyr::filter(FDR < 0.05)
  enrich_table[enrich_df$geneSet, i] <- enrich_df$enrichmentRatio
}
rownames(enrich_table) <- signif_term_df$description
```

```{r echo=FALSE}
terms_of_interest <- list()
terms_of_interest[[1]] <- 
  data.frame(description = c("axon development",
                             "central nervous system neuron differentiation",
                             "neuron projection guidance",
                             "regulation of neuron projection development"))

terms_of_interest[[2]] <- 
  data.frame(description = c("regulation of transporter activity",
                             "regulation of membrane potential",
                             "regulation of ion transmembrane transport",
                             "regulation of peptide secretion",
                             "signal release"))

terms_of_interest[[3]] <- 
  data.frame(description = c("neuropeptide signaling pathway",
                             "cell communication by electrical coupling",
                             "regulation of trans-synaptic signaling",
                             "regulation of neurological system process",
                             "neural precursor cell proliferation"))

terms_of_interest[[4]] <- 
  data.frame(description = c("RNA catabolic process",
                             "ribonucleotide metabolic process",
                             "ribonucleoprotein complex biogenesis",
                             "mitochondrial respiratory chain complex assembly",
                             "response to purine-containing compound"))

terms_of_interest_df <- do.call(rbind, terms_of_interest)
terms_of_interest_df <- left_join(terms_of_interest_df, signif_term_df,
                                  by = "description")
```

Fold of enrichment for selected GO BP terms (Figure S5E):
```{r Fig_S5E, fig.height=5.5, fig.width=8}
interest_enrich_table <- enrich_table[terms_of_interest_df$description,
                                      colnames(enrich_table) != "DYRK1A"]
interest_enrich_table[is.na(interest_enrich_table)] <- 0

map <- Heatmap(abs(interest_enrich_table),
               name = "Fold of enrichment",
               col = abs_FC_colormap,
               na_col = "grey90",
               row_title = NULL, column_title = NULL,
               cluster_rows = F, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               column_names_side = "top",
               width = unit(6, "cm"))
draw(map, heatmap_legend_side = "bottom")
```


# Session Information

```{r results="markup"}
sessionInfo()
```
